README.md
# AccuCor2 - Isotope Natural Abundance Correction for dual-isotope data

<!-- badges: start -->
[![R-CMD-check](https://github.com/wangyujue23/AccuCor2/workflows/R-CMD-check/badge.svg)](https://github.com/wangyujue23/AccuCor2/actions)
<!-- badges: end -->

AccuCor2 is an isotope natural abundance correction algorithm for dual-isotope experimental data. AccuCor2 supports 13C-2H and 13C-15N isotope correction.

AccuCor2 accepts Excel (.xls and .xlsx) files.

## Installation

Download [R](https://www.r-project.org/) to install accucor2.

```R
# Please make sure all the required packages are updated.
install.packages("devtools")
library(devtools)
devtools::install_github("wangyujue23/AccuCor2",INSTALL_opts = "--no-multiarch")
```

## Usage

```R
library(accucor2)

# Example for 13C-15N tracer data:
# Input file and the Sheet name (example file included)
# Or use your own: input_file <- "/path/to/my/datafile.xlsx"
# You may use the example files in the "inst/extdata" folder as a template to format your own data.

input_file <- system.file("extdata", "CN_Compound_Test.xlsx", package = "accucor2")
sheet_name <- "Sheet1"
metabolite_list <-system.file("extdata", "metabolite_formula_and_charge_info.csv", package = "accucor2")

# Be sure to specify the appropriate resolution.
# For Exactive, the resolution is 70000, defined at 200 Mw
# Output is written into "corrected_[time].xlsx" under the directory folder by default (a message will show you where it is)
# The results are also returned as a named list of dataframes for further processing in R
# "Original", "Corrected", "Normalized", "Pool size"

corrected <- dual_correction(input_file,sheet_name,metabolite_list,"CN",Resolution = 70000)

corrected


# Example for 13C-2H tracer data:


input_file <- system.file("extdata", "CH_Compound_Test.xlsx", package = "accucor2")
sheet_name <- "Sheet1"
metabolite_list <-system.file("extdata", "Metabolite Formula and Charge Info.csv", package = "accucor2")

# The default value for tracer C/H/N isotopic purity is 0.99, you could change them by specify the value of C13Purity and H2N15Purity.
# You could also specify the natual abundance value. In this example, the NitrogenNaturalAbundace (14N, 15N) and SulfurNaturalAbundace (32S, 33S, 34S) are specified.

corrected <- dual_correction(input_file,sheet_name,metabolite_list,"CH",C13Purity = 1, H2N15Purity = 1, Resolution = 750000,NitrogenNaturalAbundance = c(0.99632, 0.00368),SulfurNaturalAbundance = c(0.9493, 0.0076, 0.0431))

corrected

```

## Contributing
Pull requests are welcome. For major changes, please open an issue first to discuss what you would like to change.

Please make sure to update tests as appropriate.

## License

## Potential problem and solution
Error:
```R
Warning in install.packages :
  download of package ‘utf8’ failed
```
Solution:
```R
install.packages("utf8")
```

Error:
```R
Loading required package: usethis
```
Solution:
```R
install.packages("usethis")
library(usethis)
```


NEWS.md
# AccuCor2 0.1.9.9000

* Output file is now optional
* Output file format can be either `.xlsx` or `.csv`

test.R
dual_correction("Compare with real data3.xlsx","raw","KNOWN_2018May.csv","CN",Resolution = 70000)

dual_correction("CH_Compound_Test.xlsx","Sheet1","KNOWN_2018May.csv","CH",C13Purity = 1, H2N15Purity = 1, Resolution = 750000,NitrogenNaturalAbundace = c(0.99632, 0.00368),SulfurNaturalAbundace = c(0.9493, 0.0076, 0.0431))

tests/testthat.R
library(testthat)
library(accucor2)

test_check("accucor2")


tests/testthat/test_dual_correction.R
library(accucor2)

read_expected <- function(file, sheet) {
  expected <- readxl::read_excel(path = file, sheet = sheet)
  expected <- dplyr::mutate_at(
    expected,
    dplyr::vars(dplyr::ends_with("_Label")),
    as.integer
  )
}

test_that("CN correction", {
  resolution <- 70000
  input_file <- system.file("extdata",
    "CN_Compound_Test.xlsx",
    package = "accucor2"
  )
  sheet_name <- "Sheet1"
  metabolite_list <- system.file("extdata",
    "metabolite_formula_and_charge_info.csv",
    package = "accucor2"
  )

  corrected <- dual_correction(
    input_file,
    sheet_name,
    metabolite_list,
    "CN",
    Resolution = resolution,
    output_base = FALSE
  )

  expected_output <- list(
    "Original" = read_expected(
      system.file("extdata", "CN_Compound_Test_corrected.xlsx", package = "accucor2"),
      sheet = "Original"
    ),
    "Corrected" = read_expected(
      system.file("extdata", "CN_Compound_Test_corrected.xlsx", package = "accucor2"),
      sheet = "Corrected"
    ),
    "Normalized" = read_expected(
      system.file("extdata", "CN_Compound_Test_corrected.xlsx", package = "accucor2"),
      sheet = "Normalized"
    ),
    "Pool size" = read_expected(
      system.file("extdata", "CN_Compound_Test_corrected.xlsx", package = "accucor2"),
      sheet = "Pool size"
    )
  )

  expect_equal(corrected, expected_output)
})

test_that("CN correction XLSX output", {
  resolution <- 70000
  input_file <- system.file("extdata",
    "CN_Compound_Test.xlsx",
    package = "accucor2"
  )
  sheet_name <- "Sheet1"
  metabolite_list <- system.file("extdata",
    "metabolite_formula_and_charge_info.csv",
    package = "accucor2"
  )

  output_filetype <- "xlsx"
  output_file_base <-
    tempfile(fileext = paste0(".", output_filetype))
  output_file <- paste0(
    tools::file_path_sans_ext(output_file_base),
    "_corrected.",
    output_filetype
  )


  expect_message(
    dual_correction(
      input_file,
      sheet_name,
      metabolite_list,
      "CN",
      Resolution = resolution,
      output_base = output_file_base
    ),
    "Output written to:.*_corrected\\.xlsx"
  )

  expected_output_file <-
    system.file("extdata", "CN_Compound_Test_corrected.xlsx", package = "accucor2")
  expected_output <- list(
    "Original" = read_expected(expected_output_file, sheet = "Original"),
    "Corrected" = read_expected(expected_output_file, sheet = "Corrected"),
    "Normalized" = read_expected(expected_output_file, sheet = "Normalized"),
    "Pool size" = read_expected(expected_output_file, sheet = "Pool size")
  )

  corrected <- list(
    "Original" = read_expected(output_file, sheet = "Original"),
    "Corrected" = read_expected(output_file, sheet = "Corrected"),
    "Normalized" = read_expected(output_file, sheet = "Normalized"),
    "Pool size" = read_expected(output_file, sheet = "Pool size")
  )

  expect_equal(corrected, expected_output)
})


test_that("CN correction csv output", {
  resolution <- 70000
  input_file <- system.file("extdata",
    "CN_Compound_Test.xlsx",
    package = "accucor2"
  )
  sheet_name <- "Sheet1"
  metabolite_list <- system.file("extdata",
    "metabolite_formula_and_charge_info.csv",
    package = "accucor2"
  )

  output_filetype <- "csv"
  output_file_base <-
    tempfile(fileext = paste0(".", output_filetype))
  output_file <- paste0(
    tools::file_path_sans_ext(output_file_base),
    "_corrected.",
    output_filetype
  )


  expect_message(
    expect_message(
      expect_message(
        dual_correction(
          input_file,
          sheet_name,
          metabolite_list,
          "CN",
          Resolution = resolution,
          output_base = output_file_base,
          output_filetype = output_filetype
        ),
        "Output written to:.*_corrected\\.csv"
      ),
      "Output written to:.*_normalized\\.csv"
    ),
    "Output written to:.*_pool_size\\.csv"
  )

  for (sheet in list("Corrected", "Normalized", "Pool size")) {
    output_file_suffix <-
      paste0("_", sub(" ", "_", tolower(sheet)), ".", output_filetype)
    output_file <-
      paste0(
        tools::file_path_sans_ext(output_file_base),
        output_file_suffix
      )
    expect_snapshot_file(output_file,
      paste0("CN_Compound_Test", output_file_suffix),
      compare = compare_file_text
    )
  }
})



test_that("CH correction", {
  resolution <- 750000
  input_file <- system.file("extdata",
    "CH_Compound_Test.xlsx",
    package = "accucor2"
  )
  sheet_name <- "Sheet1"
  metabolite_list <- system.file("extdata",
    "metabolite_formula_and_charge_info.csv",
    package = "accucor2"
  )

  corrected <- dual_correction(
    input_file,
    sheet_name,
    metabolite_list,
    "CH",
    Resolution = resolution,
    output_base = FALSE
  )

  expected_output <- list(
    "Original" = read_expected(
      system.file("extdata", "CH_Compound_Test_corrected.xlsx", package = "accucor2"),
      sheet = "Original"
    ),
    "Corrected" = read_expected(
      system.file("extdata", "CH_Compound_Test_corrected.xlsx", package = "accucor2"),
      sheet = "Corrected"
    ),
    "Normalized" = read_expected(
      system.file("extdata", "CH_Compound_Test_corrected.xlsx", package = "accucor2"),
      sheet = "Normalized"
    ),
    "Pool size" = read_expected(
      system.file("extdata", "CH_Compound_Test_corrected.xlsx", package = "accucor2"),
      sheet = "Pool size"
    )
  )

  expect_equal(corrected, expected_output)
})

# Other possible tests
# dual_correction("Compare with real data3.xlsx","raw","KNOWN_2018May.csv","CN",Resolution = 70000)
# dual_correction("CH_Compound_Test.xlsx","Sheet1","KNOWN_2018May.csv","CH",C13Purity = 1, H2N15Purity = 1, Resolution = 750000,NitrogenNaturalAbundace = c(0.99632, 0.00368),SulfurNaturalAbundace = c(0.9493, 0.0076, 0.0431))


R/io_functions.R
write_output <-
  function(dataframe_list, path, filetype = NULL, ...) {
    if (identical(FALSE, path)) {
      return(path)
    }

    if (is.null(path)) {
      return(path)
    }

    if (is.null(filetype)) {
      filetype <- tools::file_ext(path)
    }

    if (filetype %in% c("xls", "xlsx")) {
      output_path <-
        paste(tools::file_path_sans_ext(path),
          "_corrected.",
          filetype,
          sep = ""
        )
      message(paste("Output written to: '", output_path, "'", sep = ""))
      writexl::write_xlsx(dataframe_list, output_path, ...)
    } else if (filetype %in% c("csv", "txt", "tsv")) {
      if (filetype %in% c("csv", "txt")) {
        write_func <- readr::write_csv
      } else if (filetype %in% c("tsv")) {
        write_func <- readr::write_tsv
      }
      for (sheet in names(dataframe_list)) {
        if (tolower(sheet) == "original") {
          next()
        }
        sheet_path <- paste(
          tools::file_path_sans_ext(path),
          "_",
          sub(" ", "_", tolower(sheet)),
          ".",
          filetype,
          sep = ""
        )
        message(paste("Output written to: '", sheet_path, "'", sep = ""))
        write_func(dataframe_list[[sheet]], sheet_path, ...)
      }
    } else {
      stop(paste("Unsupported output filetype: '", filetype, "'", sep = ""))
    }
    return(path)
  }


CarbonDeuteriumCorrection_V5.5.R
###Please note that this code only deals with carbon labeling experiment.
###Please use other version of correction codes for other experiments.
#Please make sure you have the packages below installed.### 
#Use install.packages("") to install packages###

require(gsubfn)
require(nnls)
require(dplyr)
require(gdata)
require(xlsx)
require(CHNOSZ)
require(stringr)

CarbonNaturalAbundace <- c(0.9893, 0.0107)
HydrogenNaturalAbundace <- c(0.999885, 0.000115)
NitrogenNaturalAbundace <- c(0.99632, 0.00368)#Used XCalibur parameter
OxygenNaturalAbundace <- c(0.99757, 0.00038, 0.00205)
SulfurNaturalAbundace <- c(0.9493, 0.0076, 0.0431)#Used XCalibur parameter
SiliconNaturalAbundace <- c(0.92223, 0.04685, 0.03092)
ChlorineNaturalAbundance<- c(0.7576,0.2424)
BromineNaturalAbundance<- c(0.5069,0.4931)

###Please make sure these parameters are accurate.
C13Purity <- 1
H2Purity <- 1
Resolution <- 750000
ResDefAt <- 200
ReportPoolSize <- TRUE
#For Exactive, the Resolution is 100000, defined at Mw 200#

#Please make sure the paths below are correct. 
#Make sure to use / in the paths

InputFile <- "CH_Compound_Test.xlsx"
InputSheetName <- "Sheet1"
MetaboliteList <- read.csv("Metabolite Formula and Charge Info.csv", header = TRUE, check.names = FALSE,stringsAsFactors=FALSE)

InputDF <- read.xlsx(InputFile, InputSheetName, header = TRUE, check.names=FALSE, stringsAsFactors=FALSE)
InputDF[,1] <- as.character(InputDF[,1])
OutputMatrix <- matrix(0, ncol=(ncol(InputDF)-5), nrow=0)
OutputPercentageMatrix <- matrix(0, ncol=(ncol(InputDF)-5), nrow=0)
OutputPoolBefore <- matrix(0, ncol=(ncol(InputDF)-5), nrow=0)
OutputPoolAfter <- matrix(0, ncol=(ncol(InputDF)-5), nrow=0)
# OutputEnrichment <- matrix(0, ncol=(ncol(InputDF)-5), nrow=0)
OutputCompound <- NULL
OutputLabel <- NULL
OutputPoolCompound <- NULL
names(InputDF)[1] <- "Compound"

if.not.null <- function(x) if(!is.null(x)) x else 0

for (i in 1:nrow(InputDF)) {
  if(startsWith(InputDF[i,1], "Unknown")) {InputDF[i,1] <- InputDF[i-1,1]}
}

Erh <- function(x) {
  n <- length(x)-1
  return(sum(x*c(0:n))/n)
}

#########################
IsotopeCorrection <- function(formula, datamatrix, label, Autofill = F, Charge = -1) {
  if (Charge == 0){
    print("Charge cannot be 0, replaced with -1")
    Charge <- -1
  }
  AtomNumber <- rep(0,9)
  names(AtomNumber) <- c("C","H","N","O","P","S","Si","Cl","Br")
  AtomicComposition <- makeup(formula)
  # adjust the H# according to the charge
  AtomicComposition["H"]<-AtomicComposition["H"]+Charge
  
  for (i in names(AtomicComposition)) {
    AtomNumber[i] <- AtomicComposition[i]
  }
  
  MZ <- sum(AtomNumber*c(12,1,14,16,31,32,28,35.5,80))/abs(Charge)
  Mass.Limit <- 1.66*MZ^1.5/Resolution/sqrt(ResDefAt)
  
  
  Mass.Difference<-(label$`13C#`)*1.00628+(label$`2H#`)*1.00335
  for (i in 1:(length(Mass.Difference)-1)){
    for (j in (i+1):length(Mass.Difference)){
      if (Mass.Limit>abs((Mass.Difference[i]-Mass.Difference[j])/Charge)){
        print(paste(compound,": 13C",label[i,1],"-2H",label[i,2]," is indistinguishable with 13C",label[j,1],"-2H",label[j,2]," under provided resolution, check your data again.",sep = ""))
      }
    }
  }
  
  CompleteLabel<-expand.grid(C13=c(0:AtomNumber["C"]),H2=c(0:AtomNumber["H"]))
  ExpMatrix <- matrix(0,ncol=ncol(datamatrix),nrow=nrow(CompleteLabel))
  LabelIndex <-c()
  for (i in 1:nrow(label)){
    for (j in 1:nrow(CompleteLabel)){
      if (prod(CompleteLabel[j,]==label[i,])){
        ExpMatrix[j,]<-datamatrix[i,]
        LabelIndex<-append(LabelIndex,j)
      }
    }
  }
  CorrectedMatrix <- matrix(0, ncol=ncol(datamatrix), nrow=nrow(CompleteLabel))
  
  if((AtomNumber["C"]+1)*(AtomNumber["H"]+1) < length(label)){
    if (length(compound)!=0){
      print(paste(compound,":the number of labeling species exceeded the maximum allowed number of carbon and hydrogen isotopomers, check the input file."))
    } else{
      print(paste(formula,":the number of labeling exceeded the number of carbon atoms, check the input file."))
    }
    return(CorrectedMatrix)
  } else{
    
    #Construct CHMatrix
    CHMatrix <- diag((AtomNumber["C"]+1)*(AtomNumber["H"]+1))
    CMatrix <- diag((AtomNumber["C"]+1))
    HMatrix <- diag((AtomNumber["H"]+1))
    for(i in 1:(AtomNumber["C"]+1)){
      CMatrix[,i] <- sapply(0:AtomNumber["C"], function(x) dbinom(x-i+1, AtomNumber["C"]-i+1 , CarbonNaturalAbundace[2]))
    }
    for(i in 1:(AtomNumber["H"]+1)){
      HMatrix[,i] <- sapply(0:AtomNumber["H"], function(x) dbinom(x-i+1, AtomNumber["H"]-i+1 , HydrogenNaturalAbundace[2]))
    }
    for(m in 1:(AtomNumber["H"]+1)){
      for (n in 1:m){
        CHMatrix[((m-1)*(AtomNumber["C"]+1)+1):(m*(AtomNumber["C"]+1)),((n-1)*(AtomNumber["C"]+1)+1):(n*(AtomNumber["C"]+1))] <- CMatrix*HMatrix[m,n]
      }
    }
    
    #Construct PurityMatrix
    PurityMatrix <- diag((AtomNumber["C"]+1)*(AtomNumber["H"]+1))
    CPurityMatrix <- diag((AtomNumber["C"]+1))
    HPurityMatrix <- diag((AtomNumber["H"]+1))
    for(i in 1:(AtomNumber["C"]+1)){
      CPurityMatrix[i,] <- sapply(0:AtomNumber["C"], function(x) dbinom(x-i+1, x , 1-C13Purity))
    }
    for(i in 1:(AtomNumber["H"]+1)){
      HPurityMatrix[i,] <- sapply(0:AtomNumber["H"], function(x) dbinom(x-i+1, x , 1-H2Purity))
    }
    for(n in 1:(AtomNumber["H"]+1)){
      for (m in 1:n){
        PurityMatrix[((m-1)*(AtomNumber["C"]+1)+1):(m*(AtomNumber["C"]+1)),((n-1)*(AtomNumber["C"]+1)+1):(n*(AtomNumber["C"]+1))] <- CPurityMatrix*HPurityMatrix[m,n]
      }
    }
    
    #Construct NonTracerMatrix  
    Isotope.Combinations <- expand.grid(C13=c(0:AtomNumber["C"]),H2=c(0:AtomNumber["H"]),N15=c(0:AtomNumber["N"]),O17=c(0:AtomNumber["O"]),
                                        O18=c(0:AtomNumber["O"]),S33=c(0:AtomNumber["S"]),S34=c(0:AtomNumber["S"]),
                                        Si29=c(0:AtomNumber["Si"]),Si30=c(0:AtomNumber["Si"]),Cl37=c(0:AtomNumber["Cl"]),
                                        Br81=c(0:AtomNumber["Br"]))
    
    Isotope.Combinations1 <- Isotope.Combinations %>% mutate(NonTracerMass=N15+O17+O18*2+S33+S34*2+Si29+Si30*2+Cl37*2+Br81*2) %>%
      filter((O17+O18)<=AtomNumber["O"] & (S33+S34)<=AtomNumber["S"] & (Si29+Si30)<=AtomNumber["Si"] & NonTracerMass<=(AtomNumber["C"]+AtomNumber["H"])) %>% filter((NonTracerMass==C13+H2))%>%
      mutate(MassDiff=0.99703*N15+1.00422*O17+2.00425*O18+0.99939*S33+1.99580*S34+0.99957*Si29+1.99684*Si30+1.99705*Cl37+1.99796*Br81
             -1.00335*C13-1.00627*H2) %>%
      filter(abs(MassDiff/Charge)<Mass.Limit)
    
    Isotope.Combinations2 <- Isotope.Combinations %>% mutate(NonTracerMass=N15+O17+O18*2+S33+S34*2+Si29+Si30*2+Cl37*2+Br81*2) %>%
      filter((O17+O18)<=AtomNumber["O"] & (S33+S34)<=AtomNumber["S"] & (Si29+Si30)<=AtomNumber["Si"] & NonTracerMass<=(AtomNumber["C"]+AtomNumber["H"])) %>% filter((NonTracerMass==abs(C13-H2))&(C13*H2*NonTracerMass>0))%>%
      mutate(MassDiff=0.99703*N15+1.00422*O17+2.00425*O18+0.99939*S33+1.99580*S34+0.99957*Si29+1.99684*Si30+1.99705*Cl37+1.99796*Br81
             -abs(1.00335*C13-1.00627*H2)) %>%
      filter(abs(MassDiff/Charge)<Mass.Limit)
    
    NonTracerMatrix <- matrix(0,ncol=(AtomNumber["C"]+1)*(AtomNumber["H"]+1),nrow=(AtomNumber["C"]+1)*(AtomNumber["H"]+1))
    for(m in 1:nrow(NonTracerMatrix)){
      Cindex<-CompleteLabel[m,1]
      Hindex<-CompleteLabel[m,2]
      Current.Combinations<-Isotope.Combinations1%>%filter(C13==Cindex & H2==Hindex)
      if (nrow(Current.Combinations)!=0){
        for (i in 1:nrow(Current.Combinations)){
          p <- dbinom(Current.Combinations[i,3], AtomNumber["N"], NitrogenNaturalAbundace[2])*
            dmultinom(unlist(c(AtomNumber["O"]-sum(Current.Combinations[i,4:5]),Current.Combinations[i,4:5])), AtomNumber["O"], OxygenNaturalAbundace)*
            dmultinom(unlist(c(AtomNumber["S"]-sum(Current.Combinations[i,6:7]),Current.Combinations[i,6:7])), AtomNumber["S"], SulfurNaturalAbundace)*
            dmultinom(unlist(c(AtomNumber["Si"]-sum(Current.Combinations[i,8:9]),Current.Combinations[i,8:9])), AtomNumber["Si"], SiliconNaturalAbundace)*
            dbinom(Current.Combinations[i,10], AtomNumber["Cl"], ChlorineNaturalAbundance[2])*
            dbinom(Current.Combinations[i,11], AtomNumber["Br"], BromineNaturalAbundance[2])
          for (n in m:nrow(NonTracerMatrix)){
            if ((n-m)%%(AtomNumber["C"]+1)<=(n-1)%%(AtomNumber["C"]+1)){
              NonTracerMatrix[n,n-m+1] <- NonTracerMatrix[n,n-m+1]+p
            }
          }
        }
      }
    }
    #Correct the situations such as 2H+15N -> 13C2
    Current.Combinations<-Isotope.Combinations2
    if (nrow(Current.Combinations)!=0){
      for (i in 1:nrow(Current.Combinations)){
        Cindex<- Current.Combinations[i,1]
        Hindex<- Current.Combinations[i,2]
        p <- dbinom(Current.Combinations[i,3], AtomNumber["N"], NitrogenNaturalAbundace[2])*
          dmultinom(unlist(c(AtomNumber["O"]-sum(Current.Combinations[i,4:5]),Current.Combinations[i,4:5])), AtomNumber["O"], OxygenNaturalAbundace)*
          dmultinom(unlist(c(AtomNumber["S"]-sum(Current.Combinations[i,6:7]),Current.Combinations[i,6:7])), AtomNumber["S"], SulfurNaturalAbundace)*
          dmultinom(unlist(c(AtomNumber["Si"]-sum(Current.Combinations[i,8:9]),Current.Combinations[i,8:9])), AtomNumber["Si"], SiliconNaturalAbundace)*
          dbinom(Current.Combinations[i,10], AtomNumber["Cl"], ChlorineNaturalAbundance[2])*
          dbinom(Current.Combinations[i,11], AtomNumber["Br"], BromineNaturalAbundance[2])
        if (Cindex<Hindex){
          n<-which(CompleteLabel$C13==Cindex&CompleteLabel$H2==0)
          m<-which(CompleteLabel$C13==0&CompleteLabel$H2==Hindex)
          for (j in 0:(AtomNumber["C"]+1-n)){
            for (h in 0: (AtomNumber["H"]-m%/%(AtomNumber["C"]+1))){
              NonTracerMatrix[m+j+h*(AtomNumber["C"]+1),n+j+h*(AtomNumber["C"]+1)] <- NonTracerMatrix[m+j+h*(AtomNumber["C"]+1),n+j+h*(AtomNumber["C"]+1)]+p
            }
          }
        } else {
          n<-which(CompleteLabel$C13==0&CompleteLabel$H2==Hindex)
          m<-which(CompleteLabel$C13==Cindex&CompleteLabel$H2==0)
          for (j in 0:(AtomNumber["C"]+1-m)){
            for (h in 0: (AtomNumber["H"]-n%/%(AtomNumber["C"]+1))){
              NonTracerMatrix[m+j+h*(AtomNumber["C"]+1),n+j+h*(AtomNumber["C"]+1)] <- NonTracerMatrix[m+j+h*(AtomNumber["C"]+1),n+j+h*(AtomNumber["C"]+1)]+p
            }
          }
        }
      }
    }
    #remove matrix rows and columns that autofilled in the input file
    if (Autofill==F){
      LabelIndex<-sort(LabelIndex)
      NonTracerMatrix<-NonTracerMatrix[LabelIndex,LabelIndex]
      CHMatrix<-CHMatrix[LabelIndex,LabelIndex]
      PurityMatrix<-PurityMatrix[LabelIndex,LabelIndex]
      ExpMatrix<-ExpMatrix[LabelIndex,]
      CorrectedMatrix<-CorrectedMatrix[LabelIndex,]
      CompleteLabel<-CompleteLabel[LabelIndex,]
    }
    
    for(i in 1:ncol(CorrectedMatrix)) {
      CorrectedMatrix[,i] <- coef(nnls(NonTracerMatrix %*% 
                                         CHMatrix %*% PurityMatrix, ExpMatrix[,i]))
    }
    CorrectedMatrix<-cbind(CompleteLabel,CorrectedMatrix)
    return(CorrectedMatrix)
  }
}

for (i in unique(InputDF$Compound)) {
  Formula=MetaboliteList[MetaboliteList$compound==i,2]
  charge=MetaboliteList[MetaboliteList$compound==i,3]
  compound <- i
  if(length(Formula)==0) {
    print(paste("The formula of",i,"is unknown",sep=" "))
    break
  }
  CurrentMetabolite <- filter(InputDF, Compound==i)
  DataMatrix <- as.matrix(CurrentMetabolite[,6:(ncol(CurrentMetabolite))])
  Corrected.raw <- IsotopeCorrection(Formula, DataMatrix, CurrentMetabolite[,3:4],Charge = charge, Autofill = F)
  Corrected<-NULL
  #Delete unexpected rows
  for (m in 1:nrow(Corrected.raw)){
    for (n in 1:nrow(CurrentMetabolite)){
      if(prod(Corrected.raw[m,1:2]==CurrentMetabolite[n,3:4]) & CurrentMetabolite[n,5] ){
        Corrected<- rbind(Corrected,Corrected.raw[m,])
      }
    }
  }
  CorrectedPercentage <- scale(Corrected[,3:ncol(Corrected)],scale=colSums(Corrected)[3:ncol(Corrected)],center=FALSE)
  OutputMatrix <- rbind(OutputMatrix, Corrected[,3:ncol(Corrected)])
  OutputPercentageMatrix <- rbind(OutputPercentageMatrix, CorrectedPercentage)
  OutputPoolBefore <- rbind(OutputPoolBefore, colSums(DataMatrix))
  OutputPoolAfter <- rbind(OutputPoolAfter, colSums(Corrected[,3:ncol(Corrected)]))
  # OutputEnrichment <- rbind(OutputEnrichment, apply(CorrectedPercentage,2,Erh))
  OutputCompound <- append(OutputCompound, rep(i,nrow(Corrected[,3:ncol(Corrected)])))
  OutputLabel <- rbind(OutputLabel, Corrected[,1:2])
  OutputPoolCompound <- append(OutputPoolCompound, i)
} 

OutputDF <- data.frame(OutputCompound, OutputLabel, OutputMatrix)
OutputPercentageDF <- data.frame(OutputCompound, OutputLabel, OutputPercentageMatrix)
OutputPoolBeforeDF <- data.frame(OutputPoolCompound, OutputPoolBefore)
OutputPoolAfterDF <- data.frame(OutputPoolCompound, OutputPoolAfter)
names(OutputDF) <- c("Compound", "C13","H2", names(InputDF)[6:(length(names(InputDF)))])
names(OutputPercentageDF) <- names(OutputDF)
names(OutputPoolBeforeDF) <- c("Compound", names(InputDF)[6:(length(names(InputDF)))])
names(OutputPoolAfterDF) <- c("Compound", names(InputDF)[6:(length(names(InputDF)))])

write.xlsx2(OutputDF, file=InputFile, sheetName = "Corrected", row.names=FALSE, append=TRUE)
write.xlsx2(OutputPercentageDF, file=InputFile, sheetName = "Normalized", row.names=FALSE, append=TRUE)
write.xlsx2(OutputPoolAfterDF, file=InputFile, sheetName = "Pool Size", row.names=FALSE, append=TRUE)

